src/metaprob/autodiff.cljc

```diff
@@ -0,0 +1,247 @@
+      (make-dual-number (:tag x1)
+                        (new-f (:value x1) x2 x3)
+                        (* (df-dx1 (:value x1) x2 x3) (:derivative x1)))
+
+      (and (map? x2)
+           (or (non-dual? x1) (clojure.core/< (:tag x1) (:tag x2)))
+           (or (non-dual? x3) (clojure.core/< (:tag x3) (:tag x2))))
+      (make-dual-number (:tag x2)
+                        (new-f x1 (:value x2) x3)
+                        (* (df-dx2 x1 (:value x2) x3) (:derivative x2)))
+
+      (and (map? x3)
+           (or (non-dual? x1) (clojure.core/< (:tag x1) (:tag x3)))
+           (or (non-dual? x2) (clojure.core/< (:tag x2) (:tag x3))))
+      (make-dual-number (:tag x3)
+                        (new-f x1 x2 (:value x3))
+                        (* (df-dx3 x1 x2 (:value x3)) (:derivative x3)))
+
+      (and (map? x1) (map? x2) (= (:tag x1) (:tag x2)) (or (non-dual? x3) (clojure.core/< (:tag x3) (:tag x1))))
+      (make-dual-number (:tag x1)
+                        (new-f (:value x1) (:value x2) x3)
+                        (+ (* (df-dx1 (:value x1) (:value x2) x3) (:derivative x1))
+                           (* (df-dx2 (:value x1) (:value x2) x3) (:derivative x2))))
+
+      (and (map? x1) (map? x3) (= (:tag x1) (:tag x3)) (or (non-dual? x2) (clojure.core/< (:tag x2) (:tag x1))))
+      (make-dual-number (:tag x1)
+                        (new-f (:value x1) x2 (:value x3))
+                        (+ (* (df-dx1 (:value x1) x2 (:value x3)) (:derivative x1))
+                           (* (df-dx3 (:value x1) x2 (:value x3)) (:derivative x3))))
+
+      (and (map? x2) (map? x3) (= (:tag x2) (:tag x3)) (or (non-dual? x1) (clojure.core/< (:tag x1) (:tag x2))))
+      (make-dual-number (:tag x2)
+                        (new-f x1 (:value x2) (:value x3))
+                        (+ (* (df-dx2 x1 (:value x2) (:value x3)) (:derivative x2))
+                           (* (df-dx3 x1 (:value x2) (:value x3)) (:derivative x3))))
+
+      (and (map? x1) (map? x2) (map? x3) (= (:tag x1) (:tag x2) (:tag x3)))
+      (make-dual-number (:tag x2)
+                        (new-f (:value x1) (:value x2) (:value x3))
+                        (+ (* (df-dx1 (:value x1) (:value x2) (:value x3)) (:derivative x1))
+                           (* (df-dx2 (:value x1) (:value x2) (:value x3)) (:derivative x2))
+                           (* (df-dx3 (:value x1) (:value x2) (:value x3)) (:derivative x3)))))))
+
+;; For functions like + and *, which can take multiple arguments
+(defn lift-real-n->real [f df-dx1 df-dx2]
+  (fn [& xs]
+    (if (nil? xs)
+      (f)
+      (reduce (lift-real*real->real f df-dx1 df-dx2) xs))))
+
+;; For functions like - and /, where the first argument is special
+(defn lift-real-n+1->real [f df-dx df-dx1 df-dx2]
+  (fn [& xs]
+    (cond (empty? xs) (f)
+          (empty? (rest xs)) ((lift-real->real f df-dx) (first xs))
+          :else (reduce (lift-real*real->real f df-dx1 df-dx2) xs))))
+
+(defn value [x]
+  (if (non-dual? x) x (:value x)))
```

Comment:
I am now realizing that this should be a recursive function! Or at least, we should have a recursive version that calls `(recur (:value x))` to unwrap `x` "all the way down" to a number.